searx/plugins/calculator/__init__.py

```diff
@@ -23,68 +22,46 @@
-    except ZeroDivisionError:
-        # This is undefined
-        return ""
-
-
-def _eval(node):
-    if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
-        return node.value
-
-    if isinstance(node, ast.BinOp):
-        return operators[type(node.op)](_eval(node.left), _eval(node.right))
-
-    if isinstance(node, ast.UnaryOp):
-        return operators[type(node.op)](_eval(node.operand))
 
-    raise TypeError(node)
-
-
-def timeout_func(timeout, func, *args, **kwargs):
-
-    def handler(q: Queue, func, args, **kwargs):  # pylint:disable=invalid-name
+def call_calculator(query_py_formatted, timeout):
+    calculator_process_py_path = Path(__file__).parent.absolute() / "calculator_process.py"
+    # see https://docs.python.org/3/using/cmdline.html
+    # -S Disable the import of the module site and the site-dependent manipulations
+    #    of sys.path that it entails. Also disable these manipulations if site is
+    #    explicitly imported later (call site.main() if you want them to be triggered).
+    # -I Run Python in isolated mode. This also implies -E, -P and -s options.
+    # -E Ignore all PYTHON* environment variables, e.g. PYTHONPATH and PYTHONHOME, that might be set.
+    # -P Don’t prepend a potentially unsafe path to sys.path
+    # -s Don’t add the user site-packages directory to sys.path.
+    process = subprocess.Popen(  # pylint: disable=R1732
+        [sys.executable, "-S", "-I", calculator_process_py_path, query_py_formatted],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+    try:
+        stdout, stderr = process.communicate(timeout=timeout)
+        if process.returncode == 0 and not stderr:
+            return stdout
+        logger.debug("calculator exited with stderr %s", stderr)
+        return None
+    except subprocess.TimeoutExpired:
+        process.terminate()
         try:
-            q.put(func(*args, **kwargs))
-        except:
-            q.put(None)
-            raise
-
-    que = Queue()
-    p = Process(target=handler, args=(que, func, args), kwargs=kwargs)
-    p.start()
-    p.join(timeout=timeout)
-    ret_val = None
-    if not p.is_alive():
-        ret_val = que.get()
-    else:
-        logger.debug("terminate function after timeout is exceeded")
-        p.terminate()
-    p.join()
-    p.close()
-    return ret_val
+            # Give the process a grace period to terminate
+            process.communicate(timeout=2)
+        except subprocess.TimeoutExpired:
+            # Forcefully kill the process
+            process.kill()
+            process.communicate()
+        logger.debug("calculator terminated after timeout")
+        # Capture any remaining output
+        return None
+    finally:
+        # Ensure the process is fully cleaned up
+        if process.poll() is None:  # If still running
+            process.kill()
+            process.communicate()
```

Comment:
Could you please explain why we're not just directly calling `process.kill()` here?

Reply:
process.poll() checks if the process has exited, which might happen before.

however subprocess.run already does that:
* https://docs.python.org/3/library/subprocess.html#subprocess.run
* https://github.com/python/cpython/blob/34ded1a1a10204635cad27830fcbee2f8547e8ed/Lib/subprocess.py#L512

I have pushed a commit to use it